### PR TITLE
Embedded priming fix for Issue #624, with Tests

### PR DESCRIPTION
**Pull request by gistya**
This is a fix for the issue where dot syntax would not work in field-names to prime. 
https://github.com/doctrine/mongodb-odm/issues/624

With this fix, now you can prime a reference that is in an embedded document. It works no matter how many levels of embedded documents there are between the original document and the final reference, but it does not traverse across multiple references. 

The recursive method is based loosely upon the structure of the prepareQueryElement() method of Doctrine\ODM\MongoDB\Persisters\DocumentPersister, except with priming it can be much simpler since we only care if it's a reference or embedded document.

### CODE REVIEW

File: lib/Doctrine/ODM/MongoDB/Query/ReferencePrimer.php [hunk 1]
```diff
@@ -157,6 +164,60 @@ public function primeReferences(ClassMetadata $class, $documents, $fieldName, ar
...
+     * This addresses Issue #624.
+     *
+     * @param array $fieldName
+     * @param ClassMetadata $class
+     * @param array\Traversable $documents
+     * @param array $mapping
+     */
+    private function parseDotSyntaxForPrimer($fieldName, $class, $documents, $mapping = null) {
+        // Recursion passthrough:
+        if($mapping != null) return array('fieldName'=>$fieldName, 'class'=>$class, 'documents'=>$documents, 'mapping'=>$mapping);
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Query/ReferencePrimer.php]: One-liners must have curly braces (later in the code as well) and look like normal ``if``s

File: lib/Doctrine/ODM/MongoDB/Query/ReferencePrimer.php [hunk 2]
```diff
@@ -157,6 +164,60 @@ public function primeReferences(ClassMetadata $class, $documents, $fieldName, ar
...
+            if(!isset($mapping['embedded']))
+                throw new \InvalidArgumentException("Field $e[0] of fieldName $fieldName is not an embedded document, therefore no children can be primed. Aborting. This feature does not support traversing nested referenced documents at this time.");
+            if(!$mapping['embedded'])
+                throw new \InvalidArgumentException("Field $e[0] of fieldName $fieldName is not an embedded document, therefore no children can be primed. Aborting. This feature does not support traversing nested referenced documents at this time.");
+            if ( ! isset($mapping['targetDocument']))
+                throw new \InvalidArgumentException("No target document class has been specified for this embedded document. However, targetDocument mapping must be specified in order for prime to work on fieldName $fieldName for mapping of field ".$mapping['fieldName']);
+            $childDocuments = array();
+            foreach($documents as $document) {
+                $fieldValue = $class->getFieldValue($document, $e[0]);
+                if(is_a($fieldValue,'Doctrine\ODM\MongoDB\PersistentCollection'))
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Query/ReferencePrimer.php]: Why not ``instanceof``?

File: lib/Doctrine/ODM/MongoDB/Query/ReferencePrimer.php [hunk 3]
```diff
@@ -157,6 +164,60 @@ public function primeReferences(ClassMetadata $class, $documents, $fieldName, ar
...
+                    foreach($fieldValue as $elemDocument) array_push($childDocuments,$elemDocument);
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Query/ReferencePrimer.php]: This also needs curly braces

File: lib/Doctrine/ODM/MongoDB/Query/ReferencePrimer.php [hunk 4]
```diff
@@ -157,6 +164,60 @@ public function primeReferences(ClassMetadata $class, $documents, $fieldName, ar
...
+     * I.e. you can prime this: myDocument.embeddedDocument.embeddedDocuments.embeddedDocuments.referencedDocument(s)
+     * ... but you cannot prime this: myDocument.embeddedDocument.referencedDocuments.referencedDocument(s)
+     * This addresses Issue #624.
+     *
+     * @param array $fieldName
+     * @param ClassMetadata $class
+     * @param array\Traversable $documents
+     * @param array $mapping
+     */
+    private function parseDotSyntaxForPrimer($fieldName, $class, $documents, $mapping = null) {
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Query/ReferencePrimer.php]: ``{`` should be in new line like everywhere else

**Comment by robholmes** [comment]
@malarzm Agreed needs a bit of a cleanup to adhere to the code standards. Is there anyway I can do this or does it need to be done by @gistya ? I guess I would need to do a new PR with with these changes in, is that right?

File: lib/Doctrine/ODM/MongoDB/Query/ReferencePrimer.php [hunk 5]
```diff
@@ -157,6 +164,60 @@ public function primeReferences(ClassMetadata $class, $documents, $fieldName, ar
...
+        if ( ! isset($class->fieldMappings[$e[0]]))
+            throw new \InvalidArgumentException("Field $fieldName cannot be further parsed for priming because it is unmpaped..");
+        $mapping = $class->fieldMappings[$e[0]];
+        $e[0] = $mapping['name'];
+
+        // Case of embedded document(s) to recurse through:
+        if(!isset($mapping['reference'])) {
+            if(!isset($mapping['embedded']))
+                throw new \InvalidArgumentException("Field $e[0] of fieldName $fieldName is not an embedded document, therefore no children can be primed. Aborting. This feature does not support traversing nested referenced documents at this time.");
+            if(!$mapping['embedded'])
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Query/ReferencePrimer.php]: This ``if`` and previous one could be merged to one ``if (empty($mapping['embedded']))``

File: lib/Doctrine/ODM/MongoDB/Query/ReferencePrimer.php [hunk 6]
```diff
@@ -157,6 +164,60 @@ public function primeReferences(ClassMetadata $class, $documents, $fieldName, ar
...
+                throw new \InvalidArgumentException("Field $e[0] of fieldName $fieldName is not an embedded document, therefore no children can be primed. Aborting. This feature does not support traversing nested referenced documents at this time.");
+            if ( ! isset($mapping['targetDocument']))
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Query/ReferencePrimer.php]: What about discriminated embedded documents? Or shall it be added later in another PR?

File: lib/Doctrine/ODM/MongoDB/Query/ReferencePrimer.php [hunk 7]
```diff
@@ -157,6 +164,60 @@ public function primeReferences(ClassMetadata $class, $documents, $fieldName, ar
...
+     * @param array $mapping
+     */
+    private function parseDotSyntaxForPrimer($fieldName, $class, $documents, $mapping = null) {
+        // Recursion passthrough:
+        if($mapping != null) return array('fieldName'=>$fieldName, 'class'=>$class, 'documents'=>$documents, 'mapping'=>$mapping);
+
+        // Gather mapping data:
+        $e = explode('.', $fieldName);
+        if ( ! isset($class->fieldMappings[$e[0]]))
+            throw new \InvalidArgumentException("Field $fieldName cannot be further parsed for priming because it is unmpaped..");
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Query/ReferencePrimer.php]: There two dots at the end of sentence

**Comment by malarzm** [comment]
@robholmes you can submit PR to @gistya's fork (to proper branch of course) and after he merges it commits will be included in this PR. Also you can import/cherry pick his commit into your branch and continue the work but that will require new PR in ODM AFAIK

**Comment by robholmes** [comment]
Thanks @malarzm I will try and get this done asap.

**Comment by jwage** [comment]
:+1:

**Comment by gistya** [comment]
This was my first pull request ever. It makes me feel good to see it get accepted. I'm glad I could help out and contribute to this awesome project.

**Comment by jwage** [comment]
thanks @gistya

**Pull request merged**